diff --git a/node_modules/mineflayer/lib/plugins/blocks.js b/node_modules/mineflayer/lib/plugins/blocks.js
index f128f71..8de5da0 100644
--- a/node_modules/mineflayer/lib/plugins/blocks.js
+++ b/node_modules/mineflayer/lib/plugins/blocks.js
@@ -200,6 +200,61 @@ function inject (bot, { version, storageBuilder, hideErrors }) {
     return blocks
   }
 
+  bot.findBlocksAsync = async (options) => {
+    const matcher = getMatchingFunction(options.matching)
+    const point = (options.point || bot.entity.position).floored()
+    const maxDistance = options.maxDistance || 16
+    const count = options.count || 1
+    const useExtraInfo = options.useExtraInfo || false
+    const yieldEvery = options.yieldEvery || 16
+    const fullMatcher = getFullMatchingFunction(matcher, useExtraInfo)
+    const start = new Vec3(Math.floor(point.x / 16), Math.floor(point.y / 16), Math.floor(point.z / 16))
+    const it = new OctahedronIterator(start, Math.ceil((maxDistance + 8) / 16))
+    const visitedSections = new Set()
+
+    let blocks = []
+    let startedLayer = 0
+    let next = start
+    let sectionsProcessed = 0
+    while (next) {
+      const column = bot.world.getColumn(next.x, next.z)
+      const sectionY = next.y + Math.abs(bot.game.minY >> 4)
+      const totalSections = bot.game.height >> 4
+      if (sectionY >= 0 && sectionY < totalSections && column && !visitedSections.has(next.toString())) {
+        const section = column.sections[sectionY]
+        if (useExtraInfo === true || isBlockInSection(section, matcher)) {
+          const begin = new Vec3(next.x * 16, sectionY * 16 + bot.game.minY, next.z * 16)
+          const cursor = begin.clone()
+          const end = cursor.offset(16, 16, 16)
+          for (cursor.x = begin.x; cursor.x < end.x; cursor.x++) {
+            for (cursor.y = begin.y; cursor.y < end.y; cursor.y++) {
+              for (cursor.z = begin.z; cursor.z < end.z; cursor.z++) {
+                if (fullMatcher(cursor) && cursor.distanceTo(point) <= maxDistance) blocks.push(cursor.clone())
+              }
+            }
+          }
+        }
+        visitedSections.add(next.toString())
+        sectionsProcessed++
+        if (sectionsProcessed % yieldEvery === 0) {
+          await new Promise(resolve => setImmediate(resolve))
+        }
+      }
+      if (startedLayer !== it.apothem && blocks.length >= count) {
+        break
+      }
+      startedLayer = it.apothem
+      next = it.next()
+    }
+    blocks.sort((a, b) => {
+      return a.distanceTo(point) - b.distanceTo(point)
+    })
+    if (blocks.length > count) {
+      blocks = blocks.slice(0, count)
+    }
+    return blocks
+  }
+
   function findBlock (options) {
     const blocks = bot.findBlocks(options)
     if (blocks.length === 0) return null
diff --git a/node_modules/mineflayer/lib/plugins/digging.js b/node_modules/mineflayer/lib/plugins/digging.js
index 496cf63..813a667 100644
--- a/node_modules/mineflayer/lib/plugins/digging.js
+++ b/node_modules/mineflayer/lib/plugins/digging.js
@@ -242,11 +242,14 @@ function inject (bot) {
       enchantments = enchantments.concat(helmetEnchantments)
     }
 
+    if (block.material === 'incorrect_for_wooden_tool')
+      block.material = 'mineable/pickaxe'
+
     const creative = bot.game.gameMode === 'creative'
     return block.digTime(
       type,
       creative,
-      bot.entity.isInWater,
+      false, // isInWater, water calc is bugged, so always false
       !bot.entity.onGround,
       enchantments,
       bot.entity.effects
diff --git a/node_modules/mineflayer/lib/plugins/inventory.js b/node_modules/mineflayer/lib/plugins/inventory.js
index 70db312..93d2054 100644
--- a/node_modules/mineflayer/lib/plugins/inventory.js
+++ b/node_modules/mineflayer/lib/plugins/inventory.js
@@ -125,10 +125,15 @@ function inject (bot, { hideErrors }) {
         cursorZ: -1
       })
     } else if (bot.supportFeature('useItemWithOwnPacket')) {
+      const yawDeg = -bot.entity.yaw * 180 / Math.PI + 180 // invert + shift
+      const pitchDeg = -bot.entity.pitch * 180 / Math.PI // invert
       bot._client.write('use_item', {
         hand: offHand ? 1 : 0,
         sequence,
-        rotation: { x: 0, y: 0 }
+        rotation: {
+          x: yawDeg,
+          y: pitchDeg
+        }
       })
     }
   }
diff --git a/node_modules/mineflayer/lib/plugins/place_block.js b/node_modules/mineflayer/lib/plugins/place_block.js
index fdaec6b..08983b6 100644
--- a/node_modules/mineflayer/lib/plugins/place_block.js
+++ b/node_modules/mineflayer/lib/plugins/place_block.js
@@ -11,7 +11,7 @@ function inject (bot) {
     let newBlock = bot.blockAt(dest)
     if (oldBlock.type === newBlock.type) {
       [oldBlock, newBlock] = await onceWithCleanup(bot, `blockUpdate:${dest}`, {
-        timeout: 5000,
+        timeout: 500,
         // Condition to wait to receive block update actually changing the block type, in case the bot receives block updates with no changes
         // oldBlock and newBlock will both be null when the world unloads
         checkCondition: (oldBlock, newBlock) => !oldBlock || !newBlock || oldBlock.type !== newBlock.type
