// mineflayer-statemachine/lib/behaviors/behaviorPlaceBlock.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BehaviorPlaceBlock = void 0;
/**
 * This behavior will attempt to place the target item against the block at the target
 * position and given target block face. If the block could not be placed for any
 * reason, this behavior fails silently.
 *
 * Even if the block could not be placed, the target item is still equipped if possible.
 */
class BehaviorPlaceBlock {
    /**
       * Creates a new mine block behavior.
       *
       * @param bot - The bot preforming the mining function.
       * @param targets - The bot targets objects.
       */
    constructor(bot, targets) {
        this.stateName = 'placeBlock';
        this.active = false;
        this.bot = bot;
        this.targets = targets;
    }
    async onStateEntered() {
        if (this.targets.item == null) {
            console.log('BehaviorPlaceBlock: item is null');
            return;
        }

        try {
            await this.bot.equip(this.targets.item, 'hand');
            console.log('Successfully equipped item:', this.targets.item.name);
        } catch (err) {
            console.log('Error equipping item:', err);
            return;
        }

        if (this.targets.position == null) {
            console.log('BehaviorPlaceBlock: position is null');
            return;
        }

        if (this.targets.blockFace == null) {
            console.log('BehaviorPlaceBlock: blockFace is null');
            return;
        }
        const block = this.bot.blockAt(this.targets.position);
        if (block == null || !this.bot.canSeeBlock(block)) {
            console.log('BehaviorPlaceBlock: block is null or cannot see block');
            return;
        }
        this.bot.placeBlock(block, this.targets.blockFace).catch(err => {
            console.log(err);
        });
    }
}
exports.BehaviorPlaceBlock = BehaviorPlaceBlock;
